### DISCUSSION TITLE
Help to test STM32WLx examples with RAK3172 Module.

### DISCUSSION THREAD

**Discussion by mmcArg** [jgromes/RadioLib#863] [Q&A]
Hi!!!!! first of all sorry for my english.
I want to test the STM32WLx examples of RadioLib with some RAK3172 Modules.
I have dubts about configuration of RF SWITCH:

in examples:

static const uint32_t rfswitch_pins[] =
                         {PC3,  PC4,  PC5};
static const Module::RfSwitchMode_t rfswitch_table[] = {
  {STM32WLx::MODE_IDLE,  {LOW,  LOW,  LOW}},
  {STM32WLx::MODE_RX,    {HIGH, HIGH, LOW}},
  {STM32WLx::MODE_TX_LP, {HIGH, HIGH, HIGH}},
  {STM32WLx::MODE_TX_HP, {HIGH, LOW,  HIGH}},
  END_OF_MODE_TABLE,
};

Looking in RAK FORUM (https://forum.rakwireless.com/t/rak3172-internal-schematic/4557) about internal module schematic;

PC13: RF SWITCH CTRL2
PB8: RF SWITCH CTRL1

someone did test RAK3172 with RadioLib? How I must configure rfswitch pins and mode table for that?
Regards!!!!!

**Comment by jgromes** [comment]
The RF switch configuration shown in the example is just that - an example. The configuration will be different for different boards, exactly as the note above it says.

You will need to figure whether the modules uses the high power (HP), low power (LP) or both amplifiers, and what configuration of PC13/PB8 switches controls whic mode (RX, TX etc.). Feel free to search the discussion and issues here, I think RAK modules were mentioned by couple people at some point.

**Reply by mikechristiansen** [reply]
@mmcArg did you get the rak3172 working with radiolib? I have super low power output and think it is my RF switch config. 

static const uint32_t rfswitch_pins[5] = {PB8, PC13, RADIOLIB_NC, RADIOLIB_NC, RADIOLIB_NC};
static const Module::RfSwitchMode_t rfswitch_table[] = {
    {STM32WLx::MODE_IDLE,  {LOW, LOW}},
    {STM32WLx::MODE_RX,    {HIGH, LOW}},
    {STM32WLx::MODE_TX_LP, {LOW, HIGH}},
    {STM32WLx::MODE_TX_HP, {LOW, HIGH}},
    END_OF_MODE_TABLE
};
Does this configuration look correct? I’ve compared the LOW/HIGH logic to the STM32Cube examples and tried to match them accordingly.

Cheers, Mike.

**Reply by HeadBoffin** [reply]
> I have super low power output and think it is my RF switch config.

Does this mean it's working? OR does it mean it's doing what you expect? What is 'super low power output' in numbers?

**Reply by StevenCellist** [reply]
RAK3172 uses SX1262 which doesn't support `MODE_TX_LP`, so you must remove that entry from the table.

**Reply by mikechristiansen** [reply]
Thanks for the replys, 
Super low output from what was expected. 
I have configured the radio for 22 output and get a -100db strength on a receiver 5 meters away, and if I go any further I get nothing. 
I have removed the LP entry with the same behaviour. 
Also if I confugre the switch with some random unused pins I get the same output making me think the switch is the problem. 
RAK have confirmed I am using the correct pins. https://forum.rakwireless.com/t/developing-for-the-rak3172-chipset-in-platformio-using-arduino/14474/5
My code is here if anyone would like to take a look. https://github.com/mikechristiansen/WS-RAK3172l-fanet/blob/main/src/main.cpp

Appreciate any input. Thanks!

**Comment by radiohound** [comment]
Hi Mike,

I am using a different module, but it also uses a STM32WLE5 ic. I’m not
sure if this pertains to your issue, but I did notice that I was not
transmitting at full power, unless I added this over current protection
code after  setOutputPower:

```
// set over current protection limit to 140 mA (accepted range is 45 - 240
mA)
if (radio.setCurrentLimit(140) == RADIOLIB_ERR_INVALID_CURRENT_LIMIT) {
Serial.println(F("Selected current limit is invalid for this module!"));
while (true);
}
```


The complete code here:

https://github.com/radiohound/PicoTrack
